### ISSUE TITLE
[ENH] Global/Panel Forecasting Using Decision Trees/Decision Tree Ensembles

### ISSUE THREAD

**Issue by satya-pattnaik** [sktime/sktime#1132]
**Is your feature request related to a problem? Please describe.**
With reference to #1074 

Given a dataset in Panel format, for .e.g

| Time  | Store Id | Sales |
| :---: | :---:    | :---: |
| Jan | Store1   | 283   |
| ... | Store1   | ...   |
| Dec | Store1   | 200   |
| Jan | Store2   | 11   |
| ... | Store2   | ...   |
| Dec | Store2   | 31   |

We need a ```make_panel_reduction``` with exogenous variables which transforms the data into the following format.
The IDs identifying ```multiple panels``` will be passed as Categorical Exogenous Variable.
  

| Lag2  | Lag1 | Store Id | Target |
| :---: | :---:| :---:    | :---:  |
| 283   | 205  | Store1   | Mar Sales S1   |
| ...   | ...  | Store1   | Dec Sales S1  |
| 11    | 13   | Store2   | Mar Sales S2  |
| ...   | ...  | Store2   | Dec Sales S2  |

So we can train a **single model** on the above reduced data set and have a Panel Forecaster. The Decision Tree/GBM/RF will make the first split based on the ```Panel Id```(which will amount to maximum reduction of variance in the data), after which, to put it loosley - the case almost becomes univariate(internal to the Decision Tree), a **naive  depiction** below. (Plotting a feature importance plot for the Decision Tree should show the Panel ID column to be the most important one.)

                               
                                

<img width="525" alt="Screenshot 2021-07-13 at 6 41 40 PM" src="https://user-images.githubusercontent.com/22102468/125457642-92940815-e9ba-4b5d-8da6-ebc503c5dd72.png">

**Advantage-**
In a practical setting where we have 1000s of SKU time series to train as in Retail, a Single Model will be sufficient instead of 1000 different univariate models.

**Caveats -**
* Boosting models don't do well when it comes to capturing **trend**, this can be taken care of by using differences of the values across different lags as exogenous/ fitting a differenced series like ARIMA etc. 

*Note* - Seasonality can be captured by using Date-time indicators like Month/Day of Week etc. as Categorical Exogenous variables.

- [Literature on this.](https://www.monash.edu/business/ebs/research/publications/ebs/wp45-2020.pdf)
- https://robjhyndman.com/publications/global-forecasting/

References-
1. [Notebook by microsoft](https://github.com/microsoft/forecasting/blob/master/examples/grocery_sales/python/00_quick_start/lightgbm_single_round.ipynb)
2. [Something similar in Neural Nets by GluonTS](https://github.com/awslabs/gluon-ts/blob/master/examples/m5_gluonts_template.ipynb)

**Describe the solution you'd like**
A clear and concise description of what you want to happen, ideally taking into consideration the existing toolbox design, classes and methods. 

**Describe alternatives you've considered**

**Additional context**

**Comment by fkiraly** [comment]
great idea!

Though this is obviously dependent on a panel forecasting design, which in turn is dependent on a consolidated set of panel data container conventions... any good ideas there, @satya-pattnaik?

**Comment by fkiraly** [comment]
your format is the "long" format, yes?

**Comment by satya-pattnaik** [comment]
> great idea!
> 
> Though this is obviously dependent on a panel forecasting design, which in turn is dependent on a consolidated set of panel data container conventions... any good ideas there, @satya-pattnaik?

Not sure about this, I have used pandas for such a work but that was very specific. Need to think about this. I hope I am getting you right here, when it comes to data container @fkiraly.

**Comment by satya-pattnaik** [comment]
> your format is the "long" format, yes?

Yes.

**Comment by RNKuhns** [comment]
This approach is similar to "pooling" information across observational units in panel regression (there you might have intercepts different panel IDs if you aren't using a panel specific regression approach like fixed effects). 

We may need to be careful how we treat the creating of the panel identification feature in the reduction. Some models can handle the categorical predictor, in which case the format @satya-pattnaik suggests would work, but may need this to be one-hot encoded.

**Comment by aiwalter** [comment]
just a stupid question: why can we not just have this as a multivariate (endog) forecasting problem? The `make_reduction` has probably anyway to be adjusted to handle this once we have multivariate interface. So the above example data would then just be:

| Time  | Sales_1 | Sales_2 | Sales_3 | Sales_X |
| :---: | :---:    | :---: | :---: | :---: |
| Dec | 283  | 123| 333| ... |
| Jan | 113| 313| 243 | ... |
| Feb | 312 | 233| 22 | ... |

So we would have a vector forecast with `n` endog variables where `n` is the amount of stores. Having this data shape, we can just apply the `make_reduction` and you can use any tree-based model. So we would not need a `make_panel_reduction` function but just adjust the `make_reduction` to multivariate (if this is not yet the case, it has to be done anyway). The function you would need to bring your data into the data shape I wrote above could be done with some pandas functions probably. But this could be contributed maybe, if this kind of data shape (what you called panel data) is often used? Not sure on this 🤔

**Comment by satya-pattnaik** [comment]
The format that you have shown here @aiwalter is suited for Vector Forecasting. Vector forecasting in such a scenario makes sense when there is inter-dependency between "Sales_1", "Sales_2"(might be because of competition).
A use case for this is - when forecasting sales of items inside the **"same"** category. Two items of same category but different brands might have inter-dependence(in their sales) because of promotions, discount on one affects the sales of the other(adversely).

CAT1_DELL<sub>T</sub> =  L1 * CAT1_DELL<sub>T-1</sub> + L2 * CAT1_MAC<sub>T-1</sub> + e1
CAT1_MAC<sub>T</sub> =  M1 * CAT1_MAC<sub>T-1</sub> + M2 * CAT1_DELL<sub>T-1</sub> + e2

Here the forecast is based on "joint evolution" of both the series([for example](https://www.statsmodels.org/dev/vector_ar.html)).

However, I argue that it doesn't make sense to have Vector Forecast when we are considering data across different stores/geographic locations/categories as there might not be any inter-dependence. 

For e.g in the M5 data - vector forecasting sales of items from "HOUSEHOLD" category and "FOOD" category will not be helpful.

I am not sure how to vector forecast using Decision Trees, the way that I know is the one I mentioned above(which is a proxy to vector forecasting) where we pass the **"group ids"** as categorical indicators and the lags to be grouped by the respective group ids.
Will be happy to discuss more on this.
Anything you want to add @ltsaprounis?

**Comment by satya-pattnaik** [comment]
I take the liberty to quote this from [Hyndman et al](https://www.monash.edu/business/ebs/research/publications/ebs/wp45-2020.pdf).

**"There is a large overlap between global models and multivariate forecasting models, because both work on groups of time series. In practical terms, global models are univariate time series models, they only require observations of a single time series for forecasting, they use the group for finding good predictive functions. Consequently, global models are more generally applicable because they do not require observations from multiple time series at the time of forecasting. Multivariate models work on groups that are supposed to have some form of interaction or joint evolution, global models work on any group. For example, forecasting a pandemic with a global model can benefit from time series measured at different historic periods (separated by decades), controlled experiments or even different pathogens, while a multivariate model is limited to the evolution of the current one and factors that influence it. Another example could be electricity consumption patterns, a global model could pool data from different places around the world to find a shared pattern, while it is clear that completely unrelated households at different corners of the world will not benefit from a multivariate approach that used their joint evolution."**

**Comment by ltsaprounis** [comment]
> Anything you want to add @ltsaprounis?

Thanks @satya-pattnaik! Not much, just more quotes :smile:

These methods were used by the majority of the high ranking M5 submissions. The paper [The M5 Accuracy competition: Results, findings and conclusions](https://www.researchgate.net/publication/344487258_The_M5_Accuracy_competition_Results_findings_and_conclusions) lists the findings from the accuracy comp in M5. One of the findings is **the value of cross-learning**:

_" In the previous M competitions, most of the series were uncorrelated, of a different frequency and domain, and chronologically unaligned. Therefore, although both of the top-performing submissions of M4 considered “cross-learning” from multiple series concurrently, instead of one series at a time, their approach was difficult to implement effectively in practice, and did not demonstrate the full potential of the approach. In contrast, since the M5 consisted of aligned, highly-correlated series structured in a hierarchical fashion, “cross-learning” was made much easier to apply, achieving superior results when compared to methods that were trained in a series-by-series fashion. Note that, apart from resulting in more accurate forecasts, “cross-learning” implies the use of a single model instead of multiple ones, each trained on the data of a different series, thus reducing overall computational cost and mitigating difficulties related to limited historical observations. Essentially, all top 50 performing methods in M5 utilized “cross-learning”, exploiting all the information being offered by the dataset."_ 

As mentioned above - in the case of M5 the global forecasting methods can also be seen as **hierarchical** as the product / store hierarchy was introduced in the model as categorical variables. Interested to hear @fkiraly's opinion on this and the implications on a hierarchical forecasting API!

In terms of the data container, in my custom implementation of this, I've used a long format pandas dataframe similar to the format from @satya-pattnaik  here:
> 
> We need a `make_panel_reduction` with exogenous variables which transforms the data into the following format.
> The IDs identifying `multiple panels` will be passed as Categorical Exogenous Variable.
> 
> | Lag2 | Lag1 | Store Id | Target |
> | ----- | ----- | --------- | ------- | 
> | 283 | 205 | Store1 | Mar Sales S1 | 
> | ... | ... | Store1 | Dec Sales S1 |
> | 11 | 13 | Store2 | Mar Sales S2 |
> | ... | ... | Store2 | Dec Sales S2 |

**Comment by fkiraly** [comment]
this has been implemented by @danbartl, with `make_reduction` being valid in the global case (instead of adding a different, `make_global_reduction`)
